rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto

```diff
@@ -2138,6 +2139,73 @@ message AdvanceTargetVersionRequest {
 // The response to a request to advance the target version of the SNS.
 message AdvanceTargetVersionResponse {}
 
+// Represents a single entry in the upgrade journal.
+// Uses `oneof` to define the type of journal entry.
+message UpgradeJournalEntry {
+  oneof entry_type {
+    // Entry for when upgrade steps are refreshed.
+    UpgradeStepsRefresh upgrade_steps_refresh = 1;
+
+    // Entry for when the target version is set.
+    TargetVersionSet target_version_set = 2;
+
+    // Entry for when an upgrade is started.
+    UpgradeStarted upgrade_started = 3;
+
+    // Entry for when an upgrade is completed.
+    UpgradeCompleted upgrade_completed = 4;
+  }
+
+  // Seconds since unix epoch that this event occured
+  optional uint64 time = 5;
+
+  // Entry type for refreshing upgrade steps.
+  message UpgradeStepsRefresh {
+    // The versions involved in the refresh.
+    optional Governance.Versions versions = 2;
+  }
+
+  // Entry type for setting the target version.
+  message TargetVersionSet {
+    // The target version set for the upgrade.
+    optional Governance.Version version = 2;
+  }
+
+  // Entry type for when an upgrade starts.
+  message UpgradeStarted {
+    // The current version before the upgrade starts.
+    optional Governance.Version current_version = 2;
+
+    // The expected version after the upgrade completes.
+    optional Governance.Version expected_version = 3;
+  }
+
+  // Entry type for when an upgrade completes.
+  message UpgradeCompleted {
+    // The status of the completed upgrade.
+    oneof status {
+      // Upgrade completed successfully.
+      Empty succeeded = 1;
+
+      // Upgrade timed out.
+      Empty timed_out = 2;
+
+      // Upgrade ended in a weird or unexpected state.
+      Governance.Version invalid_state = 3;
+
+      // Upgrade failed due to a canister call failure.
+      Empty canister_call_failure = 4;
+    }
+  }
```

Comment:
@aterga This wasn't part of the original design we discussed, but I think it may make more sense to have a completed upgrade being a separate event from an initiated upgrade - this means we can always just append to the journal instead of having to go in and edit previous entries